language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/toml/common/completion/TomlCompletionUtil.java

```diff
@@ -0,0 +1,233 @@
+            eg: [build-options],[package]
+         */
+        addTopLevelNodeCompletions(snippets.keySet(), completions);
+        return removeExistingTableKeys(completions, documentNode);
+    }
+
+    /**
+     * Returns the set of  completion items for the given set of toml nodes.
+     *
+     * @param topLevelNodes The set of top-level toml nodes.
+     * @param completions   The set of completion items.
+     * @return
+     */
+    public static Map<String, CompletionItem> addTopLevelNodeCompletions(Set<TomlNode> topLevelNodes,
+                                                                            Map<String, CompletionItem> completions) {
+        for (TomlNode topLevelNode : topLevelNodes) {
+            String key = topLevelNode.getKey();
+            if (completions.containsKey(key)) {
+                continue;
+            }
+            CompletionItem item;
+            if (topLevelNode.type() == TomlNodeType.TABLE || topLevelNode.type() == TomlNodeType.TABLE_ARRAY) {
+                item = createTopLevelCompletionItem(topLevelNode);
+            } else {
+                continue;
+            }
+            completions.put(key, item);
+        }
+        return completions;
+    }
+
+    /**
+     * Removes already declared Table completion items from the completion item list.
+     *
+     * @param completions  The set completion items
+     * @param documentNode Toml document node.
+     * @return {@link Map<String,CompletionItem>} The optimized completion item list.
+     */
+    public static Map<String, CompletionItem> removeExistingTableKeys(Map<String, CompletionItem> completions,
+                                                                      DocumentNode documentNode) {
+        List<String> existingKeys = completions.entrySet().stream()
+                .filter(entry -> isTableDeclared(documentNode, entry.getKey()))
+                .map(Map.Entry::getKey).collect(Collectors.toList());
+        for (String key : existingKeys) {
+            completions.remove(key);
+        }
+        return completions;
+    }
+
+    /**
+     * Returns if the given table key is already declared in the Toml document context.
+     *
+     * @param documentNode Toml document node.
+     * @param tableKey     table key.
+     * @return {@link Boolean}
+     */
+    public static Boolean isTableDeclared(DocumentNode documentNode, String tableKey) {
+        NodeList<DocumentMemberDeclarationNode> members = documentNode.members();
+        for (DocumentMemberDeclarationNode documentMemberNode : members) {
+            if (documentMemberNode.kind() == SyntaxKind.TABLE) {
+                TableNode rootTableNode = (TableNode) documentMemberNode;
+                String rootTableNodeKey = TomlSyntaxTreeUtil.toQualifiedName(rootTableNode.identifier().value());
+                if (tableKey.equals(rootTableNodeKey)) {
+                    return true;
+                }
```

Comment:
Shall we use early return here?

Reply:
Moved the logic to the only caller of this function